File: agent_service_agreements/negotiation.py
"""Multi-round negotiation protocol for Agent Service Agreements.

Supports propose, counter, accept, reject flows with configurable
round limits and fairness constraints.
"""

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .schema import (
    Identity,
    NegotiationMessage,
    NEGOTIATION_ACTIONS,
    _now_iso,
    _uuid,
)
from .agreement import Agreement


@dataclass
class NegotiationConfig:
    """Configuration for a negotiation session."""
    max_rounds: int = 5
    asymmetry_limit_pct: float = 25.0
    price_bound_low_multiplier: float = 0.5
    price_bound_high_multiplier: float = 3.0
    timeout_seconds: int = 3600

    def to_dict(self) -> Dict[str, Any]:
        return {
            "max_rounds": self.max_rounds,
            "asymmetry_limit_pct": self.asymmetry_limit_pct,
            "price_bound_low_multiplier": self.price_bound_low_multiplier,
            "price_bound_high_multiplier": self.price_bound_high_multiplier,
            "timeout_seconds": self.timeout_seconds,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "NegotiationConfig":
        return cls(
            max_rounds=d.get("max_rounds", 5),
            asymmetry_limit_pct=d.get("asymmetry_limit_pct", 25.0),
            price_bound_low_multiplier=d.get("price_bound_low_multiplier", 0.5),
            price_bound_high_multiplier=d.get("price_bound_high_multiplier", 3.0),
            timeout_seconds=d.get("timeout_seconds", 3600),
        )


@dataclass
class NegotiationSession:
    """A bounded negotiation between client and provider.

    Flow: PROPOSE -> (COUNTER)* -> ACCEPT | REJECT

    Each round, one party proposes changes to the agreement terms.
    Fairness constraints prevent exploitative swings.
    """

    session_id: str = ""
    agreement: Optional[Agreement] = None
    config: NegotiationConfig = field(default_factory=NegotiationConfig)
    messages: List[NegotiationMessage] = field(default_factory=list)
    current_round: int = 0
    status: str = "open"  # open, accepted, rejected, expired
    started_at: str = ""

    def __post_init__(self) -> None:
        if not self.session_id:
            self.session_id = f"neg-{_uuid()[:12]}"
        if not self.started_at:
            self.started_at = _now_iso()

    @property
    def is_open(self) -> bool:
        return self.status == "open"

    def propose(
        self,
        sender: Identity,
        agreement: Agreement,
    ) -> NegotiationMessage:
        """Initial proposal to start negotiation.

        Args:
            sender: The proposing party (typically the client)
            agreement: The proposed agreement terms
        """
        if self.messages:
            raise ValueError("Proposal already exists; use counter() for subsequent rounds")

        self.agreement = agreement
        self.agreement.status = "negotiating"

        msg = NegotiationMessage(
            negotiation_id=self.session_id,
            agreement_id=agreement.agreement_id,
            round=0,
            action="propose",
            sender=sender,
        )
        msg.compute_hash()
        self.messages.append(msg)
        return msg

    def counter(
        self,
        sender: Identity,
        proposed_changes: Dict[str, Any],
        rationale_code: str = "",
    ) -> NegotiationMessage:
        """Submit a counter-proposal.

        Args:
            sender: The counter-proposing party
            proposed_changes: Dict of field paths to new values
            rationale_code: Machine-readable rationale
        """
        if not self.is_open:
            raise ValueError(f"Negotiation is {self.status}, cannot counter")
        if not self.messages:
            raise ValueError("No proposal yet; use propose() first")

        self.current_round += 1

        if self.current_round > self.config.max_rounds:
            self.status = "rejected"
            raise ValueError(
                f"Maximum rounds ({self.config.max_rounds}) exceeded"
            )

        # Validate asymmetry constraints
        violations = self._check_asymmetry(proposed_changes)

        msg = NegotiationMessage(
            negotiation_id=self.session_id,
            agreement_id=self.agreement.agreement_id if self.agreement else "",
            round=self.current_round,
            action="counter",
            sender=sender,
            proposed_changes=proposed_changes,
            rationale_code=rationale_code,
        )
        msg.compute_hash()
        self.messages.append(msg)

        # Apply changes to the agreement
        if self.agreement:
            self._apply_changes(proposed_changes)

        return msg

    def accept(self, sender: Identity) -> NegotiationMessage:
        """Accept the current terms."""
        if not self.is_open:
            raise ValueError(f"Negotiation is {self.status}, cannot accept")
        if not self.messages:
            raise ValueError("No proposal to accept")

        self.current_round += 1
        msg = NegotiationMessage(
            negotiation_id=self.session_id,
            agreement_id=self.agreement.agreement_id if self.agreement else "",
            round=self.current_round,
            action="accept",
            sender=sender,
        )
        msg.compute_hash()
        self.messages.append(msg)
        self.status = "accepted"

        if self.agreement:
            self.agreement.status = "proposed"  # Ready for signing

        return msg

    def reject(self, sender: Identity, rationale_code: str = "") -> NegotiationMessage:
        """Reject the negotiation."""
        if not self.is_open:
            raise ValueError(f"Negotiation is {self.status}, cannot reject")

        self.current_round += 1
        msg = NegotiationMessage(
            negotiation_id=self.session_id,
            agreement_id=self.agreement.agreement_id if self.agreement else "",
            round=self.current_round,
            action="reject",
            sender=sender,
            rationale_code=rationale_code,
        )
        msg.compute_hash()
        self.messages.append(msg)
        self.status = "rejected"

        if self.agreement:
            self.agreement.status = "rejected"

        return msg

    def _check_asymmetry(self, proposed_changes: Dict[str, Any]) -> List[str]:
        """Check if proposed changes violate asymmetry limits.

        Returns list of violated field paths.
        """
        violations: List[str] = []
        limit = self.config.asymmetry_limit_pct / 100.0

        for path, new_value in proposed_changes.items():
            old_value = self._get_current_value(path)
            if old_value is None or not isinstance(old_value, (int, float)):
                continue
            if not isinstance(new_value, (int, float)):
                continue
            if old_value == 0:
                continue

            change_pct = abs(new_value - old_value) / abs(old_value)
            if change_pct > limit:
                violations.append(path)

        return violations

    def _get_current_value(self, path: str) -> Any:
        """Get a value from the current agreement by dotted path."""
        if not self.agreement:
            return None
        d = self.agreement.to_dict()
        parts = path.split(".")
        current: Any = d
        for part in parts:
            if isinstance(current, dict):
                current = current.get(part)
            elif isinstance(current, list):
                try:
                    idx = int(part)
                    current = current[idx]
                except (ValueError, IndexError):
                    return None
            else:
                return None
        return current

    def _apply_changes(self, changes: Dict[str, Any]) -> None:
        """Apply proposed changes to the agreement.

        Handles common negotiation fields: SLO values, payment amount,
        duration, and composite threshold.
        """
        if not self.agreement:
            return

        for path, value in changes.items():
            # Handle common paths
            if path.startswith("quality_criteria.dimensions["):
                self._apply_dimension_change(path, value)
            elif path == "quality_criteria.composite_threshold":
                if self.agreement.quality_criteria:
                    self.agreement.quality_criteria.composite_threshold = float(value)
            elif path == "escrow.payment.amount":
                if self.agreement.escrow:
                    self.agreement.escrow.amount = str(value)
            elif path == "service.constraints.max_duration_seconds":
                if self.agreement.service:
                    self.agreement.service.max_duration_seconds = int(value)
            elif path == "service.constraints.max_cost_usd":
                if self.agreement.service:
                    self.agreement.service.max_cost_usd = float(value)

    def _apply_dimension_change(self, path: str, value: Any) -> None:
        """Apply a change to a quality dimension's SLO."""
        if not self.agreement or not self.agreement.quality_criteria:
            return

        # Parse "quality_criteria.dimensions[0].slo.value"
        import re
        m = re.search(r"dimensions\[(\d+)\]\.(.+)", path)
        if not m:
            return

        idx = int(m.group(1))
        subpath = m.group(2)
        dims = self.agreement.quality_criteria.dimensions

        if idx >= len(dims):
            return

        dim = dims[idx]
        if subpath == "slo.value" and dim.slo:
            dim.slo.value = value
        elif subpath == "weight":
            dim.weight = float(value)

    # -- Serialization --

    def to_dict(self) -> Dict[str, Any]:
        return {
            "session_id": self.session_id,
            "agreement_id": self.agreement.agreement_id if self.agreement else "",
            "config": self.config.to_dict(),
            "current_round": self.current_round,
            "status": self.status,
            "started_at": self.started_at,
            "messages": [m.to_dict() for m in self.messages],
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "NegotiationSession":
        return cls(
            session_id=d.get("session_id", ""),
            config=NegotiationConfig.from_dict(d.get("config", {})),
            current_round=d.get("current_round", 0),
            status=d.get("status", "open"),
            started_at=d.get("started_at", ""),
            messages=[NegotiationMessage.from_dict(m) for m in d.get("messages", [])],
        )

File: typescript/src/negotiation.ts
import {
  Identity,
  NegotiationMessage,
  nowIso,
  uuid,
} from "./schema";
import { Agreement } from "./agreement";

export interface NegotiationConfigDict {
  max_rounds: number;
  asymmetry_limit_pct: number;
  price_bound_low_multiplier: number;
  price_bound_high_multiplier: number;
  timeout_seconds: number;
}

export class NegotiationConfig {
  maxRounds: number;
  asymmetryLimitPct: number;
  priceBoundLowMultiplier: number;
  priceBoundHighMultiplier: number;
  timeoutSeconds: number;

  constructor(opts: {
    maxRounds?: number;
    asymmetryLimitPct?: number;
    priceBoundLowMultiplier?: number;
    priceBoundHighMultiplier?: number;
    timeoutSeconds?: number;
  } = {}) {
    this.maxRounds = opts.maxRounds ?? 5;
    this.asymmetryLimitPct = opts.asymmetryLimitPct ?? 25.0;
    this.priceBoundLowMultiplier = opts.priceBoundLowMultiplier ?? 0.5;
    this.priceBoundHighMultiplier = opts.priceBoundHighMultiplier ?? 3.0;
    this.timeoutSeconds = opts.timeoutSeconds ?? 3600;
  }

  toDict(): NegotiationConfigDict {
    return {
      max_rounds: this.maxRounds,
      asymmetry_limit_pct: this.asymmetryLimitPct,
      price_bound_low_multiplier: this.priceBoundLowMultiplier,
      price_bound_high_multiplier: this.priceBoundHighMultiplier,
      timeout_seconds: this.timeoutSeconds,
    };
  }

  static fromDict(d: Record<string, unknown>): NegotiationConfig {
    return new NegotiationConfig({
      maxRounds: (d.max_rounds as number) ?? 5,
      asymmetryLimitPct: (d.asymmetry_limit_pct as number) ?? 25.0,
      priceBoundLowMultiplier: (d.price_bound_low_multiplier as number) ?? 0.5,
      priceBoundHighMultiplier: (d.price_bound_high_multiplier as number) ?? 3.0,
      timeoutSeconds: (d.timeout_seconds as number) ?? 3600,
    });
  }
}

export class NegotiationSession {
  sessionId: string;
  agreement: Agreement | null;
  config: NegotiationConfig;
  messages: NegotiationMessage[];
  currentRound: number;
  status: string;
  startedAt: string;

  constructor(opts: {
    sessionId?: string;
    agreement?: Agreement | null;
    config?: NegotiationConfig;
    messages?: NegotiationMessage[];
    currentRound?: number;
    status?: string;
    startedAt?: string;
  } = {}) {
    this.sessionId = opts.sessionId || `neg-${uuid().slice(0, 12)}`;
    this.agreement = opts.agreement ?? null;
    this.config = opts.config || new NegotiationConfig();
    this.messages = opts.messages || [];
    this.currentRound = opts.currentRound ?? 0;
    this.status = opts.status || "open";
    this.startedAt = opts.startedAt || nowIso();
  }

  get isOpen(): boolean {
    return this.status === "open";
  }

  propose(sender: Identity, agreement: Agreement): NegotiationMessage {
    if (this.messages.length > 0) {
      throw new Error("Proposal already exists; use counter() for subsequent rounds");
    }
    this.agreement = agreement;
    this.agreement.status = "negotiating";

    const msg = new NegotiationMessage({
      negotiationId: this.sessionId,
      agreementId: agreement.agreementId,
      round: 0,
      action: "propose",
      sender,
    });
    msg.computeHash();
    this.messages.push(msg);
    return msg;
  }

  counter(
    sender: Identity,
    proposedChanges: Record<string, unknown>,
    rationaleCode = "",
  ): NegotiationMessage {
    if (!this.isOpen) {
      throw new Error(`Negotiation is ${this.status}, cannot counter`);
    }
    if (this.messages.length === 0) {
      throw new Error("No proposal yet; use propose() first");
    }

    this.currentRound += 1;

    if (this.currentRound > this.config.maxRounds) {
      this.status = "rejected";
      throw new Error(`Maximum rounds (${this.config.maxRounds}) exceeded`);
    }

    this._checkAsymmetry(proposedChanges);

    const msg = new NegotiationMessage({
      negotiationId: this.sessionId,
      agreementId: this.agreement?.agreementId || "",
      round: this.currentRound,
      action: "counter",
      sender,
      proposedChanges: proposedChanges,
      rationaleCode,
    });
    msg.computeHash();
    this.messages.push(msg);

    if (this.agreement) {
      this._applyChanges(proposedChanges);
    }

    return msg;
  }

  accept(sender: Identity): NegotiationMessage {
    if (!this.isOpen) {
      throw new Error(`Negotiation is ${this.status}, cannot accept`);
    }
    if (this.messages.length === 0) {
      throw new Error("No proposal to accept");
    }

    this.currentRound += 1;
    const msg = new NegotiationMessage({
      negotiationId: this.sessionId,
      agreementId: this.agreement?.agreementId || "",
      round: this.currentRound,
      action: "accept",
      sender,
    });
    msg.computeHash();
    this.messages.push(msg);
    this.status = "accepted";

    if (this.agreement) {
      this.agreement.status = "proposed";
    }

    return msg;
  }

  reject(sender: Identity, rationaleCode = ""): NegotiationMessage {
    if (!this.isOpen) {
      throw new Error(`Negotiation is ${this.status}, cannot reject`);
    }

    this.currentRound += 1;
    const msg = new NegotiationMessage({
      negotiationId: this.sessionId,
      agreementId: this.agreement?.agreementId || "",
      round: this.currentRound,
      action: "reject",
      sender,
      rationaleCode,
    });
    msg.computeHash();
    this.messages.push(msg);
    this.status = "rejected";

    if (this.agreement) {
      this.agreement.status = "rejected";
    }

    return msg;
  }

  private _checkAsymmetry(proposedChanges: Record<string, unknown>): string[] {
    const violations: string[] = [];
    const limit = this.config.asymmetryLimitPct / 100.0;

    for (const [path, newValue] of Object.entries(proposedChanges)) {
      const oldValue = this._getCurrentValue(path);
      if (oldValue === null || typeof oldValue !== "number") continue;
      if (typeof newValue !== "number") continue;
      if (oldValue === 0) continue;

      const changePct = Math.abs(newValue - oldValue) / Math.abs(oldValue);
      if (changePct > limit) violations.push(path);
    }
    return violations;
  }

  private _getCurrentValue(path: string): unknown {
    if (!this.agreement) return null;
    const d = this.agreement.toDict() as unknown as Record<string, unknown>;
    const parts = path.split(".");
    let current: unknown = d;
    for (const part of parts) {
      if (current && typeof current === "object" && !Array.isArray(current)) {
        current = (current as Record<string, unknown>)[part];
      } else if (Array.isArray(current)) {
        const idx = parseInt(part, 10);
        if (isNaN(idx) || idx < 0 || idx >= current.length) return null;
        current = current[idx];
      } else {
        return null;
      }
    }
    return current;
  }

  private _applyChanges(changes: Record<string, unknown>): void {
    if (!this.agreement) return;

    for (const [path, value] of Object.entries(changes)) {
      if (path.startsWith("quality_criteria.dimensions[")) {
        this._applyDimensionChange(path, value);
      } else if (path === "quality_criteria.composite_threshold") {
        if (this.agreement.qualityCriteria) {
          this.agreement.qualityCriteria.compositeThreshold = Number(value);
        }
      } else if (path === "escrow.payment.amount") {
        if (this.agreement.escrow) {
          this.agreement.escrow.amount = String(value);
        }
      } else if (path === "service.constraints.max_duration_seconds") {
        if (this.agreement.service) {
          this.agreement.service.maxDurationSeconds = Number(value);
        }
      } else if (path === "service.constraints.max_cost_usd") {
        if (this.agreement.service) {
          this.agreement.service.maxCostUsd = Number(value);
        }
      }
    }
  }

  private _applyDimensionChange(path: string, value: unknown): void {
    if (!this.agreement?.qualityCriteria) return;

    const m = path.match(/dimensions\[(\d+)]\.(.+)/);
    if (!m) return;

    const idx = parseInt(m[1], 10);
    const subpath = m[2];
    const dims = this.agreement.qualityCriteria.dimensions;
    if (idx >= dims.length) return;

    const dim = dims[idx];
    if (subpath === "slo.value" && dim.slo) {
      dim.slo.value = value;
    } else if (subpath === "weight") {
      dim.weight = Number(value);
    }
  }

  toDict(): Record<string, unknown> {
    return {
      session_id: this.sessionId,
      agreement_id: this.agreement?.agreementId || "",
      config: this.config.toDict(),
      current_round: this.currentRound,
      status: this.status,
      started_at: this.startedAt,
      messages: this.messages.map((m) => m.toDict()),
    };
  }

  static fromDict(d: Record<string, unknown>): NegotiationSession {
    return new NegotiationSession({
      sessionId: (d.session_id as string) || "",
      config: NegotiationConfig.fromDict((d.config || {}) as Record<string, unknown>),
      currentRound: (d.current_round as number) ?? 0,
      status: (d.status as string) || "open",
      startedAt: (d.started_at as string) || "",
      messages: ((d.messages || []) as Record<string, unknown>[]).map(
        (m) => NegotiationMessage.fromDict(m as any),
      ),
    });
  }
}
